cobc/field.c

```diff
@@ -3087,18 +3164,139 @@ compute_size (struct cb_field *f)
+			}
+		}
+		if (*val == cb_zero
+		 && !f->flag_internal_register
+		 && cb_default_byte == -1
+		 && ( f->storage == CB_STORAGE_WORKING
+		   || f->storage == CB_STORAGE_LOCAL)
+		 && !f->flag_sign_separate) {
+			return 1;
+		}
+		break;
+	case CB_CATEGORY_NATIONAL:
+		/* FIXME: Fall-through, but should handle national space */
+	case CB_CATEGORY_ALPHANUMERIC:
+		if (CB_LITERAL_P (*val)) {
+			const struct cb_literal *lit = CB_LITERAL (*val);
+			char *p = (char*)lit->data;
+			char *end = p + lit->size - 1;
+			if (*end == ' ') {
+				while (p < end && *p == ' ') p++;
+				if (p == end) *val = cb_space;
+			}
+		}
+		if (*val == cb_space
+		 && !f->flag_internal_register
+		 && (cb_default_byte == -1 || cb_default_byte == ' ')
+		 && ( f->storage == CB_STORAGE_WORKING
+		   || f->storage == CB_STORAGE_LOCAL)
+		 && !f->children) {
+			return 1;
+		}
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+/* validate list of simple VALUEs */
+static int
+validate_field_value_list (cb_tree values, struct cb_field* f)
+{
+	int ret = 0;
+	int ret_single;
+	int no_relevant_value = 1;
+	for (; values; values = CB_CHAIN (values)) {
+		cb_tree value = CB_VALUE (values);
+		ret_single = validate_move (value, CB_TREE (f), 1, NULL);
+		if (!ret_single) {
+			/* possible cleanup for value */
+			if (!cleanup_field_value (f, &value)) {
+				no_relevant_value = 0;
+			}
+		} else {
+			ret++;
+		}
+	}
+	if (no_relevant_value
+	 && !ret) {
+#if 0 /* this idea was nice, but we need a pointer for
+         INITIALIZE BY VALUE, so don't drop the value; otherwise
+         fails "752. run_misc.at:10849: dump feature with NULL address ..." */
+		/* if deemed that no value is necessary: drop that completely */
+		f->values = NULL;
+#endif
+	}
+	return ret;
+}
+
+/* validate if field's VALUE clause matches
+   the necesary rules;
+   remove VALUE clause if it matches the default setting
+   to improve codegen
+   recursively called for the field's sub-fields */
 static int
 validate_field_value (struct cb_field *f)
 {
+	int ret = 0;
 	if (f->values) {
 		if (f->usage != CB_USAGE_CONTROL) {
 			if (f->flag_picture_l) {
 				cb_error_x (CB_TREE (f),
 					    _("%s and %s are mutually exclusive"),
 					    _("variable-length PICTURE"), "VALUE");
 				f->values = NULL;
+				return 1;
 			} else {
-				validate_move (CB_VALUE (f->values), CB_TREE (f), 1, NULL);
+				cb_tree x = f->values;
+				int ret_single = 0;
+				if (!CB_LIST_P (x)) {
+					/* simple, single VALUE */
+					ret_single = validate_move (x, CB_TREE (f), 1, NULL);
+					if (!ret_single) {
+						/* possible cleanup for value */
+						if (cleanup_field_value (f, &x)) {
+#if 0 /* this idea was nice, but we need a pointer for
+         INITIALIZE BY VALUE, so don't drop the value; otherwise
+         fails "752. run_misc.at:10849: dump feature with NULL address ..." */
+							/* if deemed that no value is necessary: drop that completely */
+							f->values = NULL;
+#endif
```

Comment:
please put that on the list to be reinspected; `INITIALIZE ... BY VALUE` should only init the fields that do have a value (even if it is identical to the value we'd put initial by "init by pic" option); likely we only need to adjust the comment and drop the `#if 0 `code - but this should be done after we're finished with the merge as that part is not unlikely to have been touched later on

Reply:
Done.